### ISSUE TITLE
Apply future deadlocks if quorum cannot be reached 

### ISSUE THREAD

**Issue by tylertreat** [hashicorp/raft#498]
I believe this is a duplicate of #145 which was closed due to no activity but appears to be unresolved.

I have calls to `Apply` which sometimes occur at the same time as nodes in the cluster being stopped, e.g. in unit/integration tests. In this situation, waiting on the result from the future returned by `Apply` (`deferError.Error()`) will block forever, causing the test to deadlock. This is because `deferError` is [waiting on a response from the `errCh`](https://github.com/hashicorp/raft/blob/38cb18695192d910bd774e2442d198fdc2c691a0/future.go#L110-L114) which is never going to come.

Interestingly, a [change was added a while back](https://github.com/hashicorp/raft/pull/390) to add a `ShutdownCh` to `deferError` to allow futures to react to shutdowns. However, this appears to only be used to address a deadlock that occurred when snapshots are taken.

It seems to me like `deferError` ought to have a timeout on it, either taking its own timeout on calls to `Error()` or by respecting the timeout passed to `Apply()`.

**Comment by tylertreat** [comment]
Please keep this open.

**Comment by b0nd21** [comment]
Right now (February 2025) the issue still persists. Found that yesterday on the latest version of the library. Error case: used Apply future with Error() blocking call, then if cluster is changed while the action (for example, leader changed), error doesn’t return (but seems like it should be something like ErrLeaderChanged)

**Comment by gplusplus314** [comment]
I ran into this problem today. Hopefully I'll have some time in the near future to work on a fix, but for now, let's at least try not to duplicate efforts. I can defer a fix for a while because we have a workaround that makes sense in our particular scenario.
